sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionViewsTest.java

```diff
@@ -0,0 +1,269 @@
+    Iterable<OffsetRange> ranges = Arrays.asList(range(0, 4), range(0, 4), range(0, 4));
+
+    Map<OffsetRange, Integer> nonOverlappingRangesToNumElementsPerPosition =
+        computeOverlappingRanges(ranges);
+    assertEquals(
+        ImmutableMap.builder().put(range(0, 4), 3).build(),
+        nonOverlappingRangesToNumElementsPerPosition);
+    assertNonEmptyRangesAndPositions(ranges, nonOverlappingRangesToNumElementsPerPosition);
+  }
+
+  @Test
+  public void testMultipleOverlapsForTheSameRange() {
+    Iterable<OffsetRange> ranges =
+        Arrays.asList(
+            range(0, 4),
+            range(0, 8),
+            range(0, 12),
+            range(0, 12),
+            range(4, 12),
+            range(8, 12),
+            range(0, 4),
+            range(0, 8),
+            range(0, 12),
+            range(0, 12),
+            range(4, 12),
+            range(8, 12));
+
+    Map<OffsetRange, Integer> nonOverlappingRangesToNumElementsPerPosition =
+        computeOverlappingRanges(ranges);
+    assertEquals(
+        ImmutableMap.builder().put(range(0, 4), 8).put(range(4, 8), 8).put(range(8, 12), 8).build(),
+        nonOverlappingRangesToNumElementsPerPosition);
+    assertNonEmptyRangesAndPositions(ranges, nonOverlappingRangesToNumElementsPerPosition);
+  }
+
+  @Test
+  public void testIncreasingOverlaps() {
+    Iterable<OffsetRange> ranges =
+        Arrays.asList(range(0, 4), range(1, 5), range(2, 6), range(3, 7), range(4, 8), range(5, 9));
+
+    Map<OffsetRange, Integer> nonOverlappingRangesToNumElementsPerPosition =
+        computeOverlappingRanges(ranges);
+    assertEquals(
+        ImmutableMap.builder()
+            .put(range(0, 1), 1)
+            .put(range(1, 2), 2)
+            .put(range(2, 3), 3)
+            .put(range(3, 4), 4)
+            .put(range(4, 5), 4)
+            .put(range(5, 6), 4)
+            .put(range(6, 7), 3)
+            .put(range(7, 8), 2)
+            .put(range(8, 9), 1)
+            .build(),
+        nonOverlappingRangesToNumElementsPerPosition);
+    assertNonEmptyRangesAndPositions(ranges, nonOverlappingRangesToNumElementsPerPosition);
+  }
+
+  @Test
+  public void testNestedOverlaps() {
+    Iterable<OffsetRange> ranges =
+        Arrays.asList(range(0, 8), range(1, 7), range(2, 6), range(3, 5));
+
+    Map<OffsetRange, Integer> nonOverlappingRangesToNumElementsPerPosition =
+        computeOverlappingRanges(ranges);
+    assertEquals(
+        ImmutableMap.builder()
+            .put(range(0, 1), 1)
+            .put(range(1, 2), 2)
+            .put(range(2, 3), 3)
+            .put(range(3, 5), 4)
+            .put(range(5, 6), 3)
+            .put(range(6, 7), 2)
+            .put(range(7, 8), 1)
+            .build(),
+        nonOverlappingRangesToNumElementsPerPosition);
+    assertNonEmptyRangesAndPositions(ranges, nonOverlappingRangesToNumElementsPerPosition);
+  }
+
+  @Test
+  public void testRandomRanges() {
+    Random random = new Random(123892154890L);
```

Comment:
Where'd this number come from?

Reply:
Nowhere, just wanted a stable random number so that if it failed it was repeatable.

Reply:
Makes sense. Can you add a comment?